docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+
+Let's try a different strategy: compute the columns in an `assign` call and
+then do the group by using string column names:
+
+```{python}
+%%time
+df = (
+    df.assign(
+        month=lambda t: t.uploaded_on.dt.to_period("M").dt.to_timestamp(), # <1>
+        ext=lambda t: t.path.str.extract(r"\.([a-z0-9]+)$", 0), # <1>
+    )
+    .groupby(["month", "ext"])
+    .agg({"project_name": lambda s: list(set(s))})
+    .rename(columns={"project_name": "projects"})
+)
+df
+```
+
+1. Later we'll see that Dask doesn't support lambda functions in `assign` and
+   instead we'll create an intermediate `DataFrame` for the `assign` call.
+
+Sort the values, add a new column and do the final aggregation:
+
+```{python}
+%%time
+df = (
+    df.reset_index()  # <1>
+    .sort_values("month", ascending=False)
+    .assign(
+        ext=lambda t: t.ext.str.replace(r"cxx|cpp|cc|c|hpp|h", "C/C++", regex=True)
+        .str.replace("^f.*$", "Fortran", regex=True)
+        .str.replace("rs", "Rust")
+        .str.replace("go", "Go")
+        .str.replace("asm", "Assembly")
+        .replace("", None)
+    )
+    .groupby(["month", "ext"])
+    .agg({"projects": lambda s: len(set(sum(s, [])))})
+)
+df
+```
+
+1. This call to `reset_index()` is necessary to avoid jumping through hoops to
+   do computation on a `MultiIndex` level. This is one of the trade-offs
+   associated with exposing indexes to the user--here as the keys of
+   a grouped aggregation: **grouping keys require a completely different API if
+   used in subsequent computations**.
+
+```{python}
+#| echo: false
+#
+# clean up after ourselves or else Dask will cause the kernel die from
+# running out of memory
+import gc
+
+del df
+gc.collect();  # semicolon to avoid printing the output
+```
+
+Remember, all of the previous code is executing on **a single file** and is
+taking close to three minutes to run.
+
+#### Conclusion
+
+If I only have pandas at my disposal, I'm unsure of how I can avoid getting
+a bigger computer to run this query over the entire data set.
+
+### Rewriting the query to be fair
+
+At this point I wondered whether this was a fair query to compare Dask and
+Pandas against.
+
+After all, Pandas' (and therefore Dask's) use of object arrays and their
+performance problems are well-known.
+
+The original query is using a lot of nested array types, which are very
+performant in DuckDB, but after looking at the query **we're throwing away all
+of our arrays** and we don't need to use them.
+
+Additionally, I'm using lambda functions instead of taking advantage of Pandas'
+fast built-in methods like `count`, `nunique` and others.
+
+Let's see if we can alter the original query to give Pandas and Dask a leg up.
+
+#### A story of two `GROUP BY`s
+
+It looks like we can remove the double `group_by` by moving the second `mutate`
+expression directly into the first `group_by` call. That's a nice feature of
+ibis: you don't have to first assign and then group like you do with pandas.
```

Comment:
You don't have to do this with pandas either

Reply:
Happy to use an alternative, but I doubt whether I do it in assign matters that much.

Reply:
Fair point, I can remove this bit, it is in fact possible to do this without assign in pandas, though inconvenient due to the way `extract` behaves

Reply:
I think you can get rid of that inconvenience if you set ``expand=False`` in ``extract``

Reply:
😂 Potato potato I guess!